File: app/core/permissions.py
"""Role permission matrix."""

from typing import TYPE_CHECKING

from app.models.users import UserRole

if TYPE_CHECKING:
    from app.models.users import User

PLATFORM_STAFF = frozenset({UserRole.SUPER_ADMIN})
WHOLE_ORG = frozenset({UserRole.OWNER, UserRole.ADMIN})
MANAGE_PEOPLE = frozenset({UserRole.OWNER, UserRole.ADMIN, UserRole.MANAGER})
VERIFY_SCORES = frozenset({UserRole.OWNER, UserRole.ADMIN, UserRole.MANAGER})


def is_platform_staff(user: "User") -> bool:
    return user.role in PLATFORM_STAFF


def sees_whole_org(user: "User") -> bool:
    return user.role in WHOLE_ORG


def can_manage_people(user: "User") -> bool:
    return user.role in MANAGE_PEOPLE


def can_verify_scores(user: "User") -> bool:
    """Operators never verify scores, including their own."""
    return user.role in VERIFY_SCORES

File: frontend/lib/permissions.ts
import type { Role } from "@/lib/api";

const MANAGE_PEOPLE: Role[] = ["owner", "admin", "manager"];
const MANAGE_API_KEYS: Role[] = ["owner"];
const VERIFY_SCORES: Role[] = ["owner", "admin", "manager"];
const CHANGE_ROLES: Role[] = ["owner", "admin"];

export const canManagePeople = (role: Role) => MANAGE_PEOPLE.includes(role);
export const canManageApiKeys = (role: Role) => MANAGE_API_KEYS.includes(role);
export const canVerifyScores = (role: Role) => VERIFY_SCORES.includes(role);
export const canChangeRoles = (role: Role) => CHANGE_ROLES.includes(role);

export const ROLE_LABEL: Record<Role, string> = {
  super_admin: "Платформа",
  owner: "Власник",
  admin: "Адміністратор",
  manager: "Керівник",
  operator: "Оператор",
};

export const INVITABLE_ROLES: Role[] = ["admin", "manager", "operator"];
